### PR TITLE
CCIP-4160 Token Transfer Tests

### PR DESCRIPTION
**Pull request by 0xsuryansh**
CCIP Integration test covering the following scenarios 
1. Send token to EOA
2. Send token to contract
3. Send 2 tokens to receiver
4. Send N tokens to contract

### CODE REVIEW

File: integration-tests/smoke/ccip_test.go [hunk 1]
```diff
@@ -168,105 +151,233 @@ func TestTokenTransfer(t *testing.T) {
...
+
+	tenTokens := new(big.Int).Mul(big.NewInt(1e18), big.NewInt(10))
+	mintAndApprove(t, e, state, srcToken, e.HomeChainSel, tenTokens)
+	mintAndApprove(t, e, state, dstToken, e.FeedChainSel, tenTokens)
 
-	balance, err := dstToken.BalanceOf(nil, state.Chains[tenv.FeedChainSel].Receiver.Address())
+	return srcToken, dstToken
+}
+
+func mintAndApprove(t *testing.T, e ccdeploy.DeployedEnv, state ccdeploy.CCIPOnChainState, token *burn_mint_erc677.BurnMintERC677, chainSel uint64, amount *big.Int) {
```
Comment (mateusz-sekara) [inline integration-tests/smoke/ccip_test.go]: I think we have a very similar helper for the USDC, it would be great to dedup these 
https://github.com/smartcontractkit/chainlink/blob/5a686748165e25c39f6e97d5c2141777440d508d/integration-tests/smoke/ccip_usdc_test.go#L254

File: integration-tests/smoke/ccip_test.go [hunk 2]
```diff
@@ -168,105 +151,233 @@ func TestTokenTransfer(t *testing.T) {
...
+			totalAmountsTransferred := make(map[common.Address]*big.Int)
+
+			for _, tokenAmount := range scenario.tokenAmounts {
+				dstToken := getDestinationToken(scenario, tenv, srcToken1, dstToken1, srcToken2, dstToken2, tokenAmount.Token)
+				require.NotNil(t, dstToken, "Destination token not found")
+
+				// Fetch initial balance
+				if _, exists := initialBalances[dstToken.Address()]; !exists {
+					balance, err := dstToken.BalanceOf(nil, scenario.receiver)
+					require.NoError(t, err)
```
Comment (mateusz-sekara) [inline integration-tests/smoke/ccip_test.go]: Similar story already for USDC, if worth the effort / makes sense we can dedup this code

File: integration-tests/smoke/ccip_test.go [hunk 3]
```diff
@@ -168,105 +151,233 @@ func TestTokenTransfer(t *testing.T) {
...
+			srcChain: tenv.HomeChainSel,
+			dstChain: tenv.FeedChainSel,
+			tokenAmounts: []router.ClientEVMTokenAmount{
+				{
+					Token:  srcToken1.Address(),
+					Amount: big.NewInt(1e18),
+				},
+			},
+			receiver: e.Chains[tenv.FeedChainSel].DeployerKey.From,
+			data:     []byte(""),
```
Comment (mateusz-sekara) [inline integration-tests/smoke/ccip_test.go]: I suggest making data as nil or empty slice to make it more explicit

File: integration-tests/smoke/ccip_test.go [hunk 4]
```diff
@@ -168,105 +151,233 @@ func TestTokenTransfer(t *testing.T) {
...
-	tx, err := srcToken.Mint(
-		e.Chains[tenv.HomeChainSel].DeployerKey,
-		e.Chains[tenv.HomeChainSel].DeployerKey.From,
-		new(big.Int).Mul(twoCoins, big.NewInt(10)),
-	)
-	require.NoError(t, err)
-	_, err = e.Chains[tenv.HomeChainSel].Confirm(tx)
-	require.NoError(t, err)
+	// Deploy and approve tokens.
+	srcToken1, dstToken1 := deployAndApproveTokens(t, tenv, state, "Token1")
```
Comment (mateusz-sekara) [inline integration-tests/smoke/ccip_test.go]: I think it was part of the ticket to also deploy some tokens in self serve manner. Right now, we deploy everything using the `DeployerKey`. We want to also verify it token transfers work properly in the self-serve deployed pool. Probably you need to create a new EOA and deploy token contracts from that

**Review by mateusz-sekara — COMMENTED** [review]
Good job, some minor comments

File: integration-tests/smoke/ccip_test.go [hunk 5]
```diff
@@ -112,21 +113,16 @@ func TestInitialDeployOnLocal(t *testing.T) {
 
 func TestTokenTransfer(t *testing.T) {
 	t.Parallel()
-	lggr := logger.TestLogger(t)
-	ctx := ccdeploy.Context(t)
-	tenv, _, _ := testsetups.NewLocalDevEnvironmentWithDefaultPrice(t, lggr)
-
+	tenv, state := setupEnvironment(t)
```
Comment (makramkd) [inline integration-tests/smoke/ccip_test.go]: Can you rebase? This has been extracted out in another PR and you can get rid of `setupEnvironment`

File: integration-tests/smoke/ccip_test.go [hunk 6]
```diff
@@ -168,105 +151,233 @@ func TestTokenTransfer(t *testing.T) {
...
+		srcChain     uint64
+		dstChain     uint64
+		tokenAmounts []router.ClientEVMTokenAmount
+		receiver     common.Address
+		data         []byte
+	}{
+		{
+			name:     "Send token to EOA",
+			srcChain: tenv.HomeChainSel,
+			dstChain: tenv.FeedChainSel,
```
Comment (makramkd) [inline integration-tests/smoke/ccip_test.go]: Recommendation: I've been explicitly creating `sourceChain` and `destChain` variables in the beginning of the test (after the `NewLocalDevEnvironmentWithDefaultPrice` call) and using those instead of HomeChainSel and FeedChainSel, would be good if we stayed consistent with this.

Comment (makramkd) [reply integration-tests/smoke/ccip_test.go]: Also far less confusing that using these selectors which could def be source/dest chains but have the additional property of being home/feed chains

File: integration-tests/smoke/ccip_test.go [hunk 7]
```diff
@@ -168,105 +151,233 @@ func TestTokenTransfer(t *testing.T) {
...
+	srcToken2, dstToken2 := deployAndApproveTokens(t, tenv, state, "Token2")
+
+	// Test scenarios are defined here
+	scenarios := []struct {
+		name         string
+		srcChain     uint64
+		dstChain     uint64
+		tokenAmounts []router.ClientEVMTokenAmount
+		receiver     common.Address
+		data         []byte
```
Comment (makramkd) [inline integration-tests/smoke/ccip_test.go]: If data is always nil for this test, do we need this in the scenario?

File: integration-tests/smoke/ccip_test.go [hunk 8]
```diff
@@ -168,105 +151,233 @@ func TestTokenTransfer(t *testing.T) {
...
-		require.Equal(t, ccdeploy.MockLinkPrice, timestampedPrice.Value)
+// Helper function to determine the destination token
+func getDestinationToken(scenario struct {
+	name         string
+	srcChain     uint64
+	dstChain     uint64
+	tokenAmounts []router.ClientEVMTokenAmount
+	receiver     common.Address
+	data         []byte
+}, tenv ccdeploy.DeployedEnv, srcToken1, dstToken1, srcToken2, dstToken2 *burn_mint_erc677.BurnMintERC677, tokenAddress common.Address) *burn_mint_erc677.BurnMintERC677 {
```
Comment (makramkd) [inline integration-tests/smoke/ccip_test.go]: Code smell to take an anonymous struct, better to just pass in the args you need (looks like srcChain, dstChain)

File: integration-tests/smoke/ccip_test.go [hunk 9]
```diff
@@ -168,105 +151,233 @@ func TestTokenTransfer(t *testing.T) {
...
+		} else if tokenAddress == dstToken2.Address() {
+			return srcToken2
+		}
 	}
+	return nil
+}
 
-	// Wait for all exec reports to land
-	ccdeploy.ConfirmExecWithSeqNrForAll(t, e, state, expectedSeqNum, startBlocks)
+func setupEnvironment(t *testing.T) (ccdeploy.DeployedEnv, ccdeploy.CCIPOnChainState) {
```
Comment (makramkd) [inline integration-tests/smoke/ccip_test.go]: Can remove (see above comment)

File: integration-tests/smoke/ccip_test.go (hunk 7 above)
Comment (0xsuryansh) [reply integration-tests/smoke/ccip_test.go]: For programmable token transfer tests data will not be nil right ?

File: integration-tests/smoke/ccip_test.go (hunk 8 above)
Comment (0xsuryansh) [reply integration-tests/smoke/ccip_test.go]: Good idea, I missed to remove this during a refactor

**Review by mateusz-sekara — DISMISSED** [review]

File: deployment/ccip/changeset/test_helpers.go [hunk 10]
```diff
@@ -1016,3 +1014,124 @@ func deployTransferTokenOneEnd(
...
+	e deployment.Environment,
+	state CCIPOnChainState,
+	owners map[uint64]*bind.TransactOpts,
+	tkMap map[uint64][]*burn_mint_erc677.BurnMintERC677,
+) {
+	for chain, tokens := range tkMap {
+		for _, token := range tokens {
+			twoCoins := new(big.Int).Mul(big.NewInt(1e18), big.NewInt(2))
+
+			owner, ok := owners[chain]
```
Comment (dimkouv) [inline deployment/ccip/changeset/test_helpers.go]: this could be moved in the outer loop.

File: deployment/ccip/changeset/test_helpers.go [hunk 11]
```diff
@@ -1016,3 +1014,124 @@ func deployTransferTokenOneEnd(
...
+	identifier := SourceDestPair{
+		SourceChainSelector: sourceChain,
+		DestChainSelector:   destChain,
+	}
+
+	startBlocks := make(map[uint64]*uint64)
+	expectedSeqNum := make(map[SourceDestPair]uint64)
+	expectedSeqNumExec := make(map[SourceDestPair][]uint64)
+
+	latesthdr, err := env.Chains[destChain].Client.HeaderByNumber(testcontext.Get(t), nil)
```
Comment (dimkouv) [inline deployment/ccip/changeset/test_helpers.go]: maybe better to pass `ctx` to this func?

File: deployment/ccip/changeset/test_helpers.go [hunk 12]
```diff
@@ -1016,3 +1014,124 @@ func deployTransferTokenOneEnd(
...
+	token common.Address,
+	receiver common.Address,
+	chain deployment.Chain,
+	expected *big.Int,
+) {
+	tokenContract, err := burn_mint_erc677.NewBurnMintERC677(token, chain.Client)
+	require.NoError(t, err)
+
+	require.Eventually(t, func() bool {
+		actualBalance, err := tokenContract.BalanceOf(&bind.CallOpts{Context: tests.Context(t)}, receiver)
```
Comment (dimkouv) [inline deployment/ccip/changeset/test_helpers.go]: maybe pass `ctx` to this func? to avoid getting a new one everyime.

File: deployment/ccip/changeset/test_helpers.go [hunk 13]
```diff
@@ -1016,3 +1014,124 @@ func deployTransferTokenOneEnd(
...
+func GetTokenBalance(
+	t *testing.T,
+	token common.Address,
+	receiver common.Address,
+	chain deployment.Chain,
+) *big.Int {
+	tokenContract, err := burn_mint_erc677.NewBurnMintERC677(token, chain.Client)
+	require.NoError(t, err)
+
+	balance, err := tokenContract.BalanceOf(&bind.CallOpts{Context: tests.Context(t)}, receiver)
```
Comment (dimkouv) [inline deployment/ccip/changeset/test_helpers.go]: similar for `ctx`

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+)
+
+func TestTokenTransfer(t *testing.T) {
+	lggr := logger.TestLogger(t)
+	config := &changeset.TestConfigs{}
+	tenv, _, _ := testsetups.NewLocalDevEnvironmentWithDefaultPrice(t, lggr, config)
+	inMemoryEnv := false
+
+	// use this if you are testing locally in memory
+	// tenv := changeset.NewMemoryEnvironmentWithJobsAndContracts(t, lggr, 2, 4, config)
```
Comment (dimkouv) [inline integration-tests/smoke/ccip/ccip_token_transfer_test.go:36]: would be better if we can do this using some env var. So we can easily switch in CI or run without touching code.

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+	srcToken, _, destToken, _, err := changeset.DeployTransferableToken(
+		lggr,
+		tenv.Env.Chains,
+		sourceChain,
+		destChain,
+		ownerSourceChain,
+		ownerDestChain,
+		state,
+		e.ExistingAddresses,
+		"OWNER_TOKEN",
```
Comment (dimkouv) [inline integration-tests/smoke/ccip/ccip_token_transfer_test.go:65]: would be nice if this were consts and be re-used in tests.

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+	selfServeSrcToken, _, selfServeDestToken, _, err := changeset.DeployTransferableToken(
+		lggr,
+		tenv.Env.Chains,
+		sourceChain,
+		destChain,
+		selfServeSrcTokenPoolDeployer,
+		selfServeDestTokenPoolDeployer,
+		state,
+		e.ExistingAddresses,
+		"SELF_SERVE_TOKEN",
```
Comment (dimkouv) [inline integration-tests/smoke/ccip/ccip_token_transfer_test.go:79]: same

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+		sourceChain: selfServeSrcTokenPoolDeployer,
+		destChain:   selfServeDestTokenPoolDeployer,
+	}, map[uint64][]*burn_mint_erc677.BurnMintERC677{
+		sourceChain: {selfServeSrcToken},
+		destChain:   {selfServeDestToken},
+	})
+
+	tinyOneCoin := new(big.Int).SetUint64(1)
+
+	// Test scenarios are defined here
```
Comment (dimkouv) [inline integration-tests/smoke/ccip/ccip_token_transfer_test.go]: comment probably not necessary, it's obvious

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+		expectedExecutionState int
+	}{
+		{
+			name:     "Send token to EOA",
+			srcChain: sourceChain,
+			dstChain: destChain,
+			tokenAmounts: []router.ClientEVMTokenAmount{
+				{
+					Token:  srcToken.Address(),
+					Amount: tinyOneCoin,
```
Comment (dimkouv) [inline integration-tests/smoke/ccip/ccip_token_transfer_test.go]: Name is a bit confusing for me `tinyOneCoin` maybe just `one` and `oneE18`, `fiveE18` etc ?

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+				},
+				{
+					Token:  selfServeDestToken.Address(),
+					Amount: tinyOneCoin,
+				},
+			},
+			receiver: state.Chains[sourceChain].Receiver.Address(),
+			expectedTokenBalances: map[common.Address]*big.Int{
+				selfServeSrcToken.Address(): new(big.Int).SetUint64(2),
```
Comment (dimkouv) [inline integration-tests/smoke/ccip/ccip_token_transfer_test.go]: can set `two` to a variable to follow same pattern with `tinyOneCoin`

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+	deployer *bind.TransactOpts,
+	chain deployment.Chain,
+	amountToFund *big.Int,
+	isInMemory bool,
+) *bind.TransactOpts {
+	key, err := crypto.GenerateKey()
+	require.NoError(t, err)
+
+	// Simulated backend sets chainID to 1337 always
+	chainID := big.NewInt(1337)
```
Comment (dimkouv) [inline integration-tests/smoke/ccip/ccip_token_transfer_test.go:209]: i believe it's similar for the commented out environment setup.

**Review by dimkouv — DISMISSED** [review]
Mostly minor comments, approving. Feel free to resolve and re-request review if you agree with any.

File: integration-tests/smoke/ccip/ccip_test.go [hunk 21]
```diff
@@ -75,134 +73,3 @@ func TestInitialDeployOnLocal(t *testing.T) {
 
 	// TODO: Apply the proposal.
 }
-
-func TestTokenTransfer(t *testing.T) {
```
Comment (b-gopalswami) [inline integration-tests/smoke/ccip/ccip_test.go:79]: Nit: As we moved out this test, the parallel count here can be reduced to one. https://github.com/smartcontractkit/chainlink/pull/15278/files#diff-c97a5a0f8e46926381c7f53223fb048419a40e9c9b8e243b3f49e15bc2e00d9eL945

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+				scenario.data,
+				scenario.expectedExecutionState,
+			)
+
+			for token, balance := range scenario.expectedTokenBalances {
+				expected := new(big.Int).Add(initialBalances[token], balance)
+				changeset.WaitForTheTokenBalance(t, token, scenario.receiver, e.Chains[scenario.dstChain], expected)
+			}
+		})
+	}
```
Comment (b-gopalswami) [inline integration-tests/smoke/ccip/ccip_token_transfer_test.go]: This approach is quite sequential. I would recommend splitting this test into multiple smaller tests and enabling parallel execution. 

A few options we could consider include:
1. Using multiple wallet addresses.
2. Breaking the test into separate segments (though this will require repeating the setup), and assigning a dedicated runner for each segment. This would reduce the overall execution time, as the message processing wait time can be parallelized across the different tests. With this approach, we could potentially achieve at least a 3x speedup.

**Review by 0xnogo — DISMISSED** [review]

File: deployment/ccip/changeset/test_helpers.go [hunk 23]
```diff
@@ -1016,3 +1014,124 @@ func deployTransferTokenOneEnd(
...
+			tx, err = token.Approve(e.Chains[chain].DeployerKey, state.Chains[chain].Router.Address(), twoCoins)
+			require.NoError(t, err)
+			_, err = e.Chains[chain].Confirm(tx)
+			require.NoError(t, err)
+		}
+	}
+}
+
+// TransferAndWaitForSuccess sends a message from sourceChain to destChain and waits for it to be executed
+func TransferAndWaitForSuccess(
```
Comment (makramkd) [inline deployment/ccip/changeset/test_helpers.go:1051]: I think we need a ticket to move all these funcs out of `changeset` and into a `cciptesthelpers` package or something, `changeset` is supposed to be the public API and shouldn't have these kinda funcs.

File: deployment/ccip/changeset/test_helpers.go [hunk 24]
```diff
@@ -1016,3 +1014,124 @@ func deployTransferTokenOneEnd(
...
+
+	// Wait for all commit reports to land.
+	ConfirmCommitForAllWithExpectedSeqNums(t, env, state, expectedSeqNum, startBlocks)
+
+	// Wait for all exec reports to land
+	states := ConfirmExecWithSeqNrsForAll(t, env, state, expectedSeqNumExec, startBlocks)
+	require.Equal(t, expectedStatus, states[identifier][msgSentEvent.SequenceNumber])
+}
+
+func WaitForTheTokenBalance(
```
Comment (makramkd) [inline deployment/ccip/changeset/test_helpers.go:1094]: nit: `WaitForTokenBalance` without the `The`

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go (hunk 22 above)
Comment (makramkd) [reply integration-tests/smoke/ccip/ccip_token_transfer_test.go]: I think we have to figure out a way to do parallel tests that doesn't mean we repeat a ton of test setup and no flakiness

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go (hunk 20 above)
Comment (makramkd) [reply integration-tests/smoke/ccip/ccip_token_transfer_test.go:209]: I don't think we should assume this chain ID here, these tests, in theory, can run on real chains.

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go (hunk 14 above)
Comment (mateusz-sekara) [reply integration-tests/smoke/ccip/ccip_token_transfer_test.go:36]: Yeah, we've covered that somewhere in the convo. There is a ticket to do it with env variable. Let's address it later

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go (hunk 15 above)
Comment (mateusz-sekara) [reply integration-tests/smoke/ccip/ccip_token_transfer_test.go:65]: This name is not relevant, to be honest, it could be named randomly because we never use that again. We use "owner_token" only as a documentation of what it does ;)

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go (hunk 18 above)
Comment (mateusz-sekara) [reply integration-tests/smoke/ccip/ccip_token_transfer_test.go]: Yeah, should be just oneCoin. I think it was copypasted from USDC which required always to pass exactly 1 because this value is hardcoded in the mocked contracts

File: integration-tests/smoke/ccip/ccip_token_transfer_test.go (hunk 20 above)
Comment (mateusz-sekara) [reply integration-tests/smoke/ccip/ccip_token_transfer_test.go:209]: This should be moved to test_setups. Right now we run it only on simulated or docker. If not simulated env is used we set proper chainID. I'm gonna work as followup to extract it somewhere

**Review by lukaszcl — APPROVED** [review]

**Review by mateusz-sekara — APPROVED** [review]
